rust/src/ikev1/detect.rs

```diff
@@ -0,0 +1,284 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// Author: Frank Honza <frank.honza@dcso.de>
+
+use crate::ikev1::ikev1::*;
+use std::ptr;
+use std::ffi::CStr;
+
+#[no_mangle]
+pub extern "C" fn rs_ikev1_state_get_exch_type(
+    tx: &mut IKEV1Transaction,
+    exch_type: *mut u32,
+) -> u8 {
+    if let Some(r) = tx.exchange_type {
+        unsafe{
+            *exch_type = r as u32;
+        }
+        return 1;
+    }
+
+    unsafe {
+        *exch_type = 0;
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ikev1_state_get_spi_initiator(
+    tx: &mut IKEV1Transaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    if let Some(ref r) = tx.spi_initiator {
+        let p = &format!("{:016x}", r);
+        if p.len() > 0 {
+            unsafe {
+                *buffer = p.as_ptr();
+                *buffer_len = p.len() as u32;
+            }
+            return 1;
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ikev1_state_get_spi_responder(
+    tx: &mut IKEV1Transaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    if let Some(ref r) = tx.spi_responder {
+        let p = &format!("{:016x}", r);
+        if p.len() > 0 {
+            unsafe {
+                *buffer = p.as_ptr();
+                *buffer_len = p.len() as u32;
+            }
+            return 1;
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ikev1_state_get_client_nonce(
+    state: &mut IKEV1State,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    if !state.client_nonce.is_empty() {
+        let p = &state.client_nonce;
+        if p.len() > 0 {
+            unsafe {
+                *buffer = p.as_ptr();
```

Comment:
Is your fear here that the content is changed from rust or from c or both?

And stupid question for implementation: wouldn't it be necessary to create a second variable on the heap and to copy the content over to survive the return out of the function? In this case it must be ensured that the allocated memory is freed, perhaps best in rust which would require a second rust function called when the variable is not used any longer from C?
Which would require a bit debugging there finding the correct location.
It's perhaps a bit error-prone to always remember that behaviour?
I've checked the other */detect.rs files but found no example.

In C an InspectionBuffer-struct is filled with a pointer to that variable which is returned to the more "inner" code of suricata, not sure where then best call the delete/rust_func.

Reply:
> Is your fear here that the content is changed from rust or from c or both?
> 

Unless you use the `Pin` trait in rust, a variable is not guaranteed to be at the same location, for ex if you update it. So, such affectation is correct if you check it is done once only (regardless of protocol correctness, you can have multiple
One possible scenario would be:
- first payload arrive, `client_nonce` is set
- C acquires a pointer to this nonce
- another packet with payload arrives, `client_nonce` is updated => C pointer now points to freed memory

Reply:
> And stupid question for implementation: wouldn't it be necessary to create a second variable on the heap and to copy the content over to survive the return out of the function? In this case it must be ensured that the allocated memory is freed, perhaps best in rust which would require a second rust function called when the variable is not used any longer from C?

That's what I was suggesting. If you look at `rust/src/common.rs`, you'll see the same for `CString` type.
Since here you may have a buffer, you cannot use the same types, but the logic is similar.
I think something like `return Box::into_raw(h) as *mut c_void;` is required to send a Box to C, and `{ Box::from_raw(ptr); }` to build the box (and free it immediatly).

> Which would require a bit debugging there finding the correct location.

What do you mean by "location"? Where to free the buffer?
It would be when you free the container C structure.
You can grep for `rs_cstring_free` to see an example.

> It's perhaps a bit error-prone to always remember that behaviour?

Indeed, you have to be very careful to do this correctly, and never forger to call functions.
That said, if I'm correct the pointer that was returned was not safe and could segfault.
`valgrind` can help a lot (also checks for memory leaks)

> I've checked the other */detect.rs files but found no example.
> 
> In C an InspectionBuffer-struct is filled with a pointer to that variable which is returned to the more "inner" code of suricata, not sure where then best call the delete/rust_func.

If you grep for `rs_krb5_tx_get_cname`, I did that for Kerberos cname matching. However, the difference in Kerberos is that the returned pointer is part of the _transaction_, and so never changes (the transaction is assigned only once). This is different for the state, which can be updated if I'm correct.
The transaction memory will be freed, so there is no leak.

So, the TL;DR is:
- if there is no way to reassign/change/free the value from Rust, you're fine (for ex. a transaction)
- or, you can create a pair of functions to allocate+copy / free memory in rust and call them from C
- or, you can allocate from C and call rust to copy data in the allocated buffer
- maybe other solutions ...

I hope it is clear!

Reply:
Thanks for your fast answer! i'll check your proposals!
Perhaps it's indeed the best to store more stuff in transactions. I had some other problem with alerts occuring more often than i expected, perhaps the problem was from the beginning that they are persisted in the "longer living" state.